[Badatprogrammiing] iP

TerrenceFyx opened the pull request:
### Duke First
> “Your mind is for having ideas, not holding them.” – David Allen [source](https://dansilvestre.com/productivity-quotes)

DukePro frees your mind of having to remember things you need to do. It's,
* text-based
* easy to learn
* ~~FAST ~~ SUPER FAST to use

All you need to do is,
1. download it from [here](https://nus-cs2103-ay2324s1.github.io/website/schedule/week4/project.html)
2. double-click it.
3. add your tasks.
4. let it manage your tasks for you 😉

And it is **FREE **!

Features:
- [x] Managing tasks
- [ ] Managing deadlines (coming soon)
- [ ] Reminders (coming soon)

***
If you Java programmer, you can use it to practice Java too. Here's the `main`method

```java
public class Main {
    public static void main(String[] args) {
        Application.launch(MainApp.class, args);
    }
}
```

Commit:
Add Gradle support

Eclipse-Dominator committed:
Bump gradle and lib version

Commit:
Add Rename, Greet, Exit features

Commit:
Add Echo feature

Commit:
Add List and Add task feature

Commit:
Added Mark, Unmark task features

Commit:
Add Todo, Event, Deadline Tasks

Commit:
Add automated ui tests

Commit:
Add error handling using exceptions

Commit:
Add delete feature

Commit:
Seperate various task creations in createTask method

Commit:
Add local save and load when tasks changes

Commit:
Remove unused imports

Commit:
Covert dates from String to LocalDate

Commit:
Add error handling for invalid dates

Commit:
Change invalid date error message

Commit:
Merge branch 'branch-Level-8'

Commit:
Refactor code into different classes

Commit:
Move code into duke package

Commit:
Merge branch 'add-gradle-support' of https://github.com/nus-cs2103-AY2324S1/ip

Commit:
Add junit test cases

Commit:
Update storage for creation of .jar

Commit:
Change invalid dead error msg

Commit:
Add JavaDocs for all public methods and classes

Commit:
Change code to follow coding standard

Commit:
Add support and test cases for find function

Commit:
Merge branch 'branch-A-JavaDoc'

Commit:
Merge branch 'branch-A-CodingStandard'

Commit:
Merge branch 'branch-Level-9'

Commit:
Fix some coding standard issues

Commit:
Add gui and minimal stying

Commit:
Add font to the gui

Commit:
Made gui app larger

Commit:
Merge branch 'branch-Level-10'

fuyiqiao commented on src/main/java/duke/CommandParser.java:
```diff
@@ -0,0 +1,128 @@
+
+  private Task createTodoTask(String[] words) {
+    StringBuilder taskName = new StringBuilder();
+    for (int i = 1; i < words.length; i += 1) {
```
using `i++` might be more concise

fuyiqiao commented on src/main/java/duke/CommandParser.java:
```diff
@@ -0,0 +1,128 @@
+        Platform.exit();
+      };
+    } else if (words[0].equals("list") && words.length == 1) {
+      return (taskList, Storage, vBox) -> {
```
Storage should be storage (first letter not capitalised)

fuyiqiao commented on src/main/java/duke/TaskList.java:
```diff
@@ -0,0 +1,155 @@
+   * @param index the position of the tasks to be unmarked
+   * @throws WrongIndexException when index is out of bounds, or too big to parse
+   */
+  public void unmarkedTask(String index, Pane pane, Image dukeImage) throws WrongIndexException {
```
`unmarkTask`

fuyiqiao reviewed (commented):
Very well organised! I like how you created your own custom errors :)

Commit:
Made gui responsive

ElginTZM commented on src/main/java/duke/CommandParser.java:
```diff
@@ -0,0 +1,128 @@
+ */
+public class CommandParser {
+
+  /**
+   * Creates a parser to parse input commands
+   */
+  public CommandParser() {}
```
May not be necessary to include an empty constructor

ElginTZM commented on src/main/java/duke/Task.java:
```diff
@@ -0,0 +1,72 @@
+    this.isCompleted = isCompleted;
+  }
+
+  public abstract String saveData();
```
Perhaps can name this method as `getSaveData` instead of `saveData` to make it clearer

ElginTZM commented on src/main/java/duke/Task.java:
```diff
@@ -0,0 +1,72 @@
+    return this.isCompleted;
+  }
+
+  protected String taskName() {
```
Similar to above, maybe can use verbs for function names?

ElginTZM commented on src/main/java/duke/Task.java:
```diff
@@ -0,0 +1,72 @@
+   *
+   * @return the current completion status
+   */
+  protected boolean isCompleted() {
```
Perhaps can use a verb for function names like `getIsCompleted` like you've done above for `markCompleted`

ElginTZM commented on src/main/java/duke/WrongIndexException.java:
```diff
@@ -0,0 +1,14 @@
+  /**
+   * Creates a WrongIndexException.
+   */
+  WrongIndexException() {
```
Could the default modifier cause an issue?

ElginTZM reviewed (commented):
Everything looks good! Code is easy to read and understand, just some minor nitpicks with how you name certain functions in `Command.java`.

Commit:
Add assertions to verify invariants, pre and post
conditions

Add assertions in TaskList, Storage, CommandParser classes.
Specifically making assertions when tasks size changes,
task being mark or unmark and control flow invariants.

Commit:
Made code adhere to code quality standard

TerrenceFyx committed:
Merge pull request #1 from Badatprogrammiing/branch-A-Assertions

Add assertions to verify invariants, pre and post conditions

Commit:
Merge branch 'master' into branch-A-CodeQuality

TerrenceFyx committed:
Merge pull request #2 from Badatprogrammiing/branch-A-CodeQuality

Made code adhere to code quality standard

Commit:
Add undo feature

Able to undo all the latest changes made in the current session.

Commit:
Update README
